File: paleta/pages/image_drop/dropped_image.py
from gi.repository import Adw, GLib, Gtk, GdkPixbuf
from PIL import Image

class DroppedImage(Adw.Bin):
    __gtype_name__ = 'DroppedImage'

    def __init__(self, image_path) -> None:
        super().__init__()
        self.set_halign(Gtk.Align.CENTER)
        self.set_valign(Gtk.Align.CENTER)
        self.load_image(image_path)

    def load_image(self, image_path):
        self.image_path = image_path
        source_image = Image.open(image_path)
        width, height = source_image.size
        bytes_ = source_image.tobytes("raw", "RGBA")
        glib_bytes = GLib.Bytes.new(bytes_)
        pixbuf = GdkPixbuf.Pixbuf.new_from_bytes(glib_bytes, GdkPixbuf.Colorspace.RGB, True, 8, width, height, width*4)

        self.picture = Gtk.Picture.new_for_pixbuf(pixbuf)
        self.set_child(self.picture)

    

File: src/pages/image_drop/dropped_image.rs
/* dropped_image.rs
 *
 * SPDX-FileCopyrightText: 2023 nate-xyz
 * SPDX-License-Identifier: GPL-3.0-or-later
 */

use adw::{prelude::*, subclass::prelude::*};
use gtk::{glib, gdk_pixbuf};

use std::{cell::RefCell, path::Path, error::Error};

mod imp {
    use super::*;

    #[derive(Debug, Default)]
    pub struct DroppedImagePriv {
        pub picture: RefCell<gtk::Picture>,
        pub pixbuf: RefCell<Option<gdk_pixbuf::Pixbuf>>,
    }

    #[glib::object_subclass]
    impl ObjectSubclass for DroppedImagePriv {
        const NAME: &'static str = "DroppedImage";
        type Type = super::DroppedImage;
        type ParentType = adw::Bin;
    }

    impl ObjectImpl for DroppedImagePriv {
        fn constructed(&self) {
            self.parent_constructed();
            self.obj().initialize();
        }
    }

    impl WidgetImpl for DroppedImagePriv {}
    impl BinImpl for DroppedImagePriv {}

    impl DroppedImagePriv {}
}

glib::wrapper! {
    pub struct DroppedImage(ObjectSubclass<imp::DroppedImagePriv>)
    @extends gtk::Widget, adw::Bin;
}


impl DroppedImage {
    pub fn new() -> DroppedImage {
        let dropped_image: DroppedImage = glib::Object::builder::<DroppedImage>().build();
        dropped_image
    }

    fn initialize(&self) {
      self.set_halign(gtk::Align::Center);
      self.set_valign(gtk::Align::Center);
    }

    pub fn load_image(&self, image_path: &str) -> Result<(), Box<dyn Error>> {
        let imp = self.imp();
        let path = Path::new(image_path);
        let basename = path.file_name().unwrap();
        let _image_name = format!("{}", html_escape::encode_text_minimal(basename.to_str().unwrap()));
        let pixbuf = gdk_pixbuf::Pixbuf::from_file(image_path)?;
        let picture = gtk::Picture::for_pixbuf(&pixbuf);
        self.set_child(Some(&picture));
        imp.picture.replace(picture);
        imp.pixbuf.replace(Some(pixbuf));
        Ok(())
    }
}
    